frontend/i_types.py

```diff
@@ -0,0 +1,274 @@
+            return False
+        for i in range(len(self.types)):
+            if not self.types[i].is_subtype(t.types[i]):
+                return False
+        return True
+
+    def get_name(self):	
+        types_names = [t.get_name() for t in self.types]
+        return "Tuple({})".format(",".join(types_names))
+
+    def get_possible_tuple_slicings(self):
+        """Returns a union type of all possible slicings of this tuple
+
+		For example:
+			t = (1, "string", 2.5)
+			
+			t.get_possible_tuple_slicings() will return the following
+			Union{Tuple(Int), Tuple(Int,String), Tuple(Int,String,Float), Tuple(String), Tuple(String,Float), Tuple(Float), Tuple()}
+						
+		"""
+        slicings = {TTuple([])}
+        for i in range(len(self.types)):
+            for j in range(i + 1, len(self.types) + 1):
+                slicings.add(TTuple(self.types[i:j]))
+        return UnionTypes(slicings)
+				
+class TIterator(Type):
+    """Type given to an iterator.
+
+    Attributes:
+        type (Type): Type of the iterator.
+    """
+
+    def __init__(self, t):
+        self.type = t
+
+    def is_subtype(self, t):
+        return isinstance(t, TIterator) and self.type.is_subtype(t.type)
+
+    def get_name(self):
+        return "Iterator({})".format(self.type.get_name())
+
+
+class TDictionary(Type):
+    """Type given to a dictionary, whose keys are of the same type, and values are of the same type.
+
+    Attributes:
+        key_type (Type): Type of the dictionary keys.
+        value_type (Type): Type of the dictionary values.
+    """
+
+    def __init__(self, t_k, t_v):
+        self.key_type = t_k
+        self.value_type = t_v
+
+    def is_subtype(self, t):
+        return (isinstance(t, TDictionary) and self.key_type.is_subtype(t.key_type)
+            and self.value_type.is_subtype(t.value_type))
+
+    def get_name(self):
+        return "Dict({}:{})".format(self.key_type.get_name(), self.value_type.get_name())
+
+
+class TSet(Type):
+    """Type given to homogeneous sets"""
+
+    def __init__(self, t):
+        self.type = t
+
+    def is_subtype(self, t):
+        return isinstance(t, TSet) and self.type.is_subtype(t.type)
+
+    def get_name(self):
+        return "Set({})".format(self.type.get_name())
+
+
+class TFunction(Type):
+    """Type given to a function.
+
+    Attributes:
+        return_type (Type): Type of the function return value.
+        arg_types ([Type]): A list of types for the function arguments.
+    """
+
+    def __init__(self, t_r, t_a):
+        self.return_type = t_r
+        self.arg_types = t_a
+
+    def is_subtype(self, t):
+        if not isinstance(t, TFunction):
+            return False
+        if len(self.arg_types) != len(t.arg_types):
+            return False
+        if not self.return_type.is_subtype(t.return_type):
+            return False
+        for i in range(len(self.arg_types)):
+            if not self.arg_types[i].is_subtype(t.arg_types[i]):
```

Comment:
This should be the other way round; if f and g are function types and f is a subtype of g, then the types of the parameters of f have to be supertypes of the types of the parameters of g.